File: frontend/src/lib/desktop/graphics-settings.ts
// SPDX-FileCopyrightText: 2026 Quad4 Software
// SPDX-License-Identifier: Apache-2.0

import type {
  GraphicsEnvironment,
  GraphicsSettings,
} from "@bindings/melovian/services/models.js";

export type { GraphicsEnvironment, GraphicsSettings };
export type NvExplicitSyncMode = "auto" | "on" | "off";

export function defaultGraphicsSettings(): GraphicsSettings {
  return {
    disableDmabufRenderer: true,
    disableCompositingMode: false,
    nvDisableExplicitSync: "auto",
  };
}

export function mergeGraphicsSettings(
  partial: Partial<GraphicsSettings> | null | undefined,
): GraphicsSettings {
  const defaults = defaultGraphicsSettings();
  if (!partial) return defaults;
  const nvMode =
    partial.nvDisableExplicitSync === "on" ||
    partial.nvDisableExplicitSync === "off" ||
    partial.nvDisableExplicitSync === "auto"
      ? partial.nvDisableExplicitSync
      : defaults.nvDisableExplicitSync;
  return {
    disableDmabufRenderer:
      typeof partial.disableDmabufRenderer === "boolean"
        ? partial.disableDmabufRenderer
        : defaults.disableDmabufRenderer,
    disableCompositingMode:
      typeof partial.disableCompositingMode === "boolean"
        ? partial.disableCompositingMode
        : defaults.disableCompositingMode,
    nvDisableExplicitSync: nvMode,
  };
}

export async function loadGraphicsEnvironment(): Promise<GraphicsEnvironment | null> {
  const { MediaService } = await import("@bindings/melovian/services/index.js");
  const env = await MediaService.GetGraphicsEnvironment();
  if (!env?.supported) return null;
  return {
    ...env,
    settings: mergeGraphicsSettings(env.settings),
    defaults: mergeGraphicsSettings(env.defaults),
    appliedEnv: env.appliedEnv ?? {},
  };
}

export async function saveGraphicsSettings(
  settings: GraphicsSettings,
): Promise<void> {
  const { MediaService } = await import("@bindings/melovian/services/index.js");
  await MediaService.SaveGraphicsSettings(mergeGraphicsSettings(settings));
}

export function nvExplicitSyncLabel(mode: NvExplicitSyncMode): string {
  switch (mode) {
    case "on":
      return "On";
    case "off":
      return "Off";
    default:
      return "Auto (Wayland + NVIDIA)";
  }
}

export function appliedEnvSummary(
  applied: Record<string, string | undefined>,
): string {
  const entries = Object.entries(applied).filter(
    (entry): entry is [string, string] => typeof entry[1] === "string",
  );
  if (entries.length === 0) return "No workarounds active in this session.";
  return entries.map(([key, value]) => `${key}=${value}`).join(", ");
}

File: internal/desktop/graphics_settings.go
// SPDX-FileCopyrightText: 2026 Quad4 Software
// SPDX-License-Identifier: Apache-2.0

package desktop

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
)

const graphicsSettingsFile = "graphics-settings.json"

// NvExplicitSyncMode controls the NVIDIA Wayland explicit-sync workaround.
type NvExplicitSyncMode string

const (
	NvExplicitSyncAuto NvExplicitSyncMode = "auto"
	NvExplicitSyncOn   NvExplicitSyncMode = "on"
	NvExplicitSyncOff  NvExplicitSyncMode = "off"
)

// GraphicsSettings stores user preferences for Linux WebKit/NVIDIA workarounds.
// Changes take effect on the next application launch.
type GraphicsSettings struct {
	DisableDmabufRenderer  bool               `json:"disableDmabufRenderer"`
	DisableCompositingMode bool               `json:"disableCompositingMode"`
	NvDisableExplicitSync  NvExplicitSyncMode `json:"nvDisableExplicitSync"`
}

// GraphicsEnvironment describes the host graphics stack for the settings UI.
type GraphicsEnvironment struct {
	Supported           bool              `json:"supported"`
	Platform            string            `json:"platform"`
	Wayland             bool              `json:"wayland"`
	Nvidia              bool              `json:"nvidia"`
	Settings            GraphicsSettings  `json:"settings"`
	Defaults            GraphicsSettings  `json:"defaults"`
	AppliedEnv          map[string]string `json:"appliedEnv"`
	RequiresRestart     bool              `json:"requiresRestart"`
	RestartRequiredNote string            `json:"restartRequiredNote,omitempty"`
}

func DefaultGraphicsSettings() GraphicsSettings {
	return GraphicsSettings{
		DisableDmabufRenderer:  true,
		DisableCompositingMode: false,
		NvDisableExplicitSync:  NvExplicitSyncAuto,
	}
}

func graphicsSettingsPath(dataDir string) string {
	return filepath.Join(dataDir, graphicsSettingsFile)
}

func LoadGraphicsSettings(dataDir string) GraphicsSettings {
	defaults := DefaultGraphicsSettings()
	raw, err := os.ReadFile(graphicsSettingsPath(dataDir))
	if err != nil {
		return defaults
	}
	var stored GraphicsSettings
	if err := json.Unmarshal(raw, &stored); err != nil {
		return defaults
	}
	return MergeGraphicsSettings(stored)
}

func MergeGraphicsSettings(partial GraphicsSettings) GraphicsSettings {
	defaults := DefaultGraphicsSettings()
	out := defaults
	out.DisableDmabufRenderer = partial.DisableDmabufRenderer
	out.DisableCompositingMode = partial.DisableCompositingMode
	switch partial.NvDisableExplicitSync {
	case NvExplicitSyncOn, NvExplicitSyncOff, NvExplicitSyncAuto:
		out.NvDisableExplicitSync = partial.NvDisableExplicitSync
	default:
		out.NvDisableExplicitSync = NvExplicitSyncAuto
	}
	return out
}

func SaveGraphicsSettings(dataDir string, settings GraphicsSettings) error {
	if dataDir == "" {
		return fmt.Errorf("data directory is required")
	}
	merged := MergeGraphicsSettings(settings)
	raw, err := json.MarshalIndent(merged, "", "  ")
	if err != nil {
		return err
	}
	path := graphicsSettingsPath(dataDir)
	if err := os.MkdirAll(filepath.Dir(path), 0o750); err != nil {
		return err
	}
	return os.WriteFile(path, raw, 0o600)
}

func nvExplicitSyncEnabled(settings GraphicsSettings, wayland, nvidia bool) bool {
	switch settings.NvDisableExplicitSync {
	case NvExplicitSyncOn:
		return true
	case NvExplicitSyncOff:
		return false
	default:
		return wayland && nvidia
	}
}
